src/main/scala/workflow/NodeOptimizationRule.scala

```diff
@@ -0,0 +1,96 @@
+          case _ => Set[Int]()
+        }
+
+      case (EstimatorFitNode(estIndex, _), index) =>
+        instructions(estIndex) match {
+          case _: OptimizableEstimator[_, _] => WorkflowUtils.getParents(index, instructions) + index
+          case _: OptimizableLabelEstimator[_, _, _] => WorkflowUtils.getParents(index, instructions) + index
+          case _ => Set[Int]()
+        }
+
+      case _ => Set[Int]()
+    }.reduce(_ union _) -- WorkflowUtils.getChildren(Pipeline.SOURCE, instructions) - Pipeline.SOURCE
+
+    // Execute the minimal amount necessary of the pipeline on sampled nodes, and optimize the optimizable nodes
+    val optimizedInstructions = instructions.toArray
+    val registers = new Array[InstructionOutput](instructions.length)
+    val numPerPartitionPerNode = Array.fill[Option[Map[Int, Int]]](instructions.length)(None)
+
+    for ((instruction, index) <- instructions.zipWithIndex) {
+      if (instructionsToExecute.contains(index)) {
+        instruction match {
+          case SourceNode(rdd) =>
+            val sampledRDD = rdd.sample(withReplacement = false, sampleFraction, seed)
+            registers(index) = RDDOutput(sampledRDD)
+            numPerPartitionPerNode(index) = Some(WorkflowUtils.numPerPartition(rdd))
+
+          case TransformerApplyNode(tIndex, inputIndices) =>
+            val numPerPartition = numPerPartitionPerNode(inputIndices.head)
+            val inputs = inputIndices.map(registers).collect {
+              case RDDOutput(rdd) => rdd.cache()
+            }
+            inputs.foreach(_.count())
+
+            // Optimize the transformer if possible
+            val transformer = registers(tIndex) match {
+              case TransformerOutput(ot: OptimizableTransformer[a, b]) =>
+                ot.optimize(inputs.head.asInstanceOf[RDD[a]], numPerPartition.get)
+              case TransformerOutput(t) => t
+              case _ => throw new ClassCastException("TransformerApplyNode dep wasn't pointing at a transformer")
+            }
+
+            registers(index) = RDDOutput(transformer.transformRDD(inputs))
+            optimizedInstructions(tIndex) = transformer
+            numPerPartitionPerNode(index) = numPerPartition
+
+          case EstimatorFitNode(estIndex, inputIndices) =>
+            val numPerPartition = numPerPartitionPerNode(inputIndices.head)
+            val inputs = inputIndices.map(registers).collect {
+              case RDDOutput(rdd) => rdd.cache()
+            }
+            inputs.foreach(_.count())
+
+            // Optimize the estimator if possible
+            val estimator = registers(estIndex) match {
+              case EstimatorOutput(oe: OptimizableEstimator[a, b]) =>
+                oe.optimize(inputs.head.asInstanceOf[RDD[a]], numPerPartition.get)
+              case EstimatorOutput(oe: OptimizableLabelEstimator[a, b, l]) =>
+                oe.optimize(inputs.head.asInstanceOf[RDD[a]], inputs(1).asInstanceOf[RDD[l]], numPerPartition.get)
```

Comment:
Style nit - is there a reason not to use `(0)` in places of `.head` here since you're using `(1)` in the same line?